### PR TITLE
chore(healthcheck): make endpoint use kafka heartbeat

### PR DESCRIPTION
**Pull request by hazzadous**
Rather than doing a full end to end healthcheck in the liveness check
(which is meant to be light), we instead check that the last heartbeat
of the consumer is within the configured `sessionTimeout`. If we are
able to call describe on the group we also check if any rebalancing is
happening that might suggest we haven't got to heartbeat stage yet.

This is still probably too complicated for a liveness check, but the
http server code path is reasonably far removed from the Kafka consumer
path so it seems possible that us serving up HTTP requests successfully
isn't a good indicator that we will be able to recover.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Review by guidoiaquinti — APPROVED** [review]
Not the best reviewer for this part of the codebase but overall LGTM

File: plugin-server/src/main/ingestion-queues/kafka-queue.ts [hunk 1]
```diff
@@ -20,18 +20,27 @@ export class KafkaQueue {
...
     private consumer: Consumer
+    private consumerConfig: Partial<ConsumerConfig>
     private consumerGroupMemberId: string | null
     private wasConsumerRan: boolean
     private ingestionTopic: string
     private eventsTopic: string
     private eachBatch: Record<string, EachBatchFunction>
 
-    constructor(pluginsServer: Hub, workerMethods: WorkerMethods) {
+    constructor(pluginsServer: Hub, workerMethods: WorkerMethods, consumerConfig: { sessionTimeout: 30000 }) {
```
Comment (macobo) [inline plugin-server/src/main/ingestion-queues/kafka-queue.ts:33]: This doesn't seem correct - you're passing `{ sessionTimeout: 30000 }` as a type and we're not overriding this anywhere.

Please put plugin-server configuration into plugin-server/src/config/config.ts and use that.

File: plugin-server/src/main/ingestion-queues/kafka-queue.ts [hunk 2]
```diff
@@ -180,6 +189,23 @@ export class KafkaQueue {
...
+        const sessionTimeout = this.consumerConfig?.sessionTimeout || 30000
+        if (this.lastHeartbeat && Date.now() - this.lastHeartbeat < sessionTimeout) {
+            return true
+        }
+
+        // Consumer has no heartbeat, but maybe it's because the group is currently rebalancing
+        try {
+            const { state } = await this.consumer.describeGroup()
+            return ['CompletingRebalance', 'PreparingRebalance'].includes(state)
+        } catch (err) {
```
Comment (macobo) [inline plugin-server/src/main/ingestion-queues/kafka-queue.ts:204]: What errors are we swallowing here? Why do we want to do so?

**Review by macobo — CHANGES_REQUESTED** [review]
A problem and potential unanswered question - both should be resolved before merging.

The approach looks fine - let's try it out.

File: plugin-server/src/main/ingestion-queues/kafka-queue.ts [hunk 3]
```diff
@@ -180,6 +189,23 @@ export class KafkaQueue {
             .some(({ topic, partitions }) => topic === targetTopic && (!partition || partitions.includes(partition)))
     }
 
+    async isHealthy(): Promise<boolean> {
```
Comment (macobo) [inline plugin-server/src/main/ingestion-queues/kafka-queue.ts:192]: If we weren't rushing this - all plugin-server code should have proper tests, including this.

File: plugin-server/src/main/ingestion-queues/kafka-queue.ts [hunk 4]
```diff
@@ -20,18 +20,27 @@ export class KafkaQueue {
...
         this.pluginsServer = pluginsServer
         this.kafka = pluginsServer.kafka!
-        this.consumer = KafkaQueue.buildConsumer(this.kafka, this.consumerGroupId())
+        this.consumerConfig = consumerConfig
+        this.consumer = KafkaQueue.buildConsumer(this.kafka, this.consumerGroupId(), consumerConfig)
+
+        // Record when the last heartbeat was, so we can check how alive the
+        // `KafkaQueue` is
+        this.lastHeartbeat = undefined
+        this.consumer.on(this.consumer.events.HEARTBEAT, ({ timestamp }) => (this.lastHeartbeat = timestamp))
```
Comment (macobo) [inline plugin-server/src/main/ingestion-queues/kafka-queue.ts:42]: Doing side-effects in constructor is in bad taste. Please move to within `start()`

**Comment by hazzadous** [comment]
Closing in favour of https://github.com/PostHog/posthog/pull/11234 at least for now. Maybe revisit if we see unrecoverable pods

**Pull request closed without merging**